### ISSUE TITLE
TypeScript to JavaScript Build

### ISSUE THREAD

**Issue by ritch** [strongloop/loopback-next#159]
Concerns:

 - we can't keep using `ts-node` it is inconsistent with the `tsc` output
 - dev experience of LB apps shouldn't suffer because of build step
 - we should run our tests against the production grade build (eg. tsc output)

Depends on #158 

- [x] Try compiling + publishing with one simple package first (@loopback/testlab)
  - [x] `npm run build` should output to `lib/*` via `tsc`
  - https://github.com/strongloop/loopback-next/pull/219
- [x] Smoke test published @loopback/testlab in an external project
  - https://github.com/strongloop-internal/scrum-asteroid/issues/126

Outcome of spike is a checklist to complete at #240

**Comment by ritch** [comment]
@bajtos @superkhau can help with this

**Comment by bajtos** [comment]
I am proposing the following setup:

 - `src/` contains original TypeScript sources
 - `lib/` contains transpiled output targeting es2017 (Node 7, 8 and newer).
  The idea is to get the benefits of native async/await on platforms that support it.
 - `lib-es6/` or perhaps `lib6` or `lib2015`contains tranpiled output targeting ES6/ES2015 (Node 6).
  The idea is to keep support for people building loopback-next applications on Node 6, which will be in LTS for the next year and then in Maintenance for another year. See [Node.js LTS schedule](https://raw.githubusercontent.com/nodejs/LTS/89730a8f9ada6e2b310566a02e3d9c5020745c28/schedule.png).
 - `index.js` reads `process.versions.node` or `process.versions.v8` and decides which `lib*` to load

(Yes, we will need to run `tsc` twice - once for `lib` and then for `lib6` again.)

I haven't formed a clear opinion on how to transpile test files. It may be ok to run our TypeScript-based tests via `ts-node` as long as we also run few high-level ES6/2017 tests via vanilla `node`.

**Comment by bajtos** [comment]
Quoting from https://github.com/strongloop/loopback-next/issues/92:

In order to continually verify the correctness of this setup, a simple end-to-end test written in pure javascript should be included.

**Comment by bajtos** [comment]
@superkhau your spike in #219 does not take into account what I proposed in my https://github.com/strongloop/loopback-next/issues/159#issuecomment-294864796.

Also your setup runs the transpiler only at prepublish time (which is run after `lerna bootstrap` too). I think that means the unit-tests in `testlab` will not pick up the changes made in `testlab`'s TS files.

I think we need to setup `tsc` in a watch mode to automatically recompile files as they are changed, this should also improve the speed of red-green-refactor cycle.

**Comment by bajtos** [comment]
>  - `lib/` contains transpiled output targeting es2017 (Node 7, 8 and newer).
  The idea is to get the benefits of native async/await on platforms that support it.
>  - `lib-es6/` or perhaps `lib6` or `lib2015`contains tranpiled output targeting ES6/ES2015 (Node 6).
  The idea is to keep support for people building loopback-next applications on Node 6, which will be in LTS for the next year and then in Maintenance for another year. See [Node.js LTS schedule](https://raw.githubusercontent.com/nodejs/LTS/89730a8f9ada6e2b310566a02e3d9c5020745c28/schedule.png).

Let me emphasise why I think this is important:
 - Starting with Node.js 7.6 (or 7.7?), `async/await` are awailable as native constructs and they received significant speed-up in the upcoming 8.0 version (scheduled for the end of May ATM).
 - The code emitted by TSC is about 4x slower than native async/await IIRC. 
 - If we ship only es6 code with async/await transpiled away, then users of the latest Node.j 8+ versions will suffer unnecessary performance penalty.

**Comment by superkhau** [comment]
@bajtos I only tried publishing to see if everything would work properly as part of the spike. I am now ready to make changes based on your suggestions.

1. I will compile for es2015
2. I will compile for 7+

Any ideas on what the root level index.js exports? Right now it's pointing at the 7+ version.

>I think we need to setup tsc in a watch mode to automatically recompile files as they are changed, this should also improve the speed of red-green-refactor cycle.

Out of scope for this issue. Although I do agree that we need some sort of watcher to recompile on the fly. My main concern is making sure we can compile/publish manually for now + make sure the acceptance criteria above is met.

**Comment by superkhau** [comment]
>Also your setup runs the transpiler only at prepublish time (which is run after lerna bootstrap too). I think that means the unit-tests in testlab will not pick up the changes made in testlab's TS files.

You have to manually recompile for now via `npm run build`.
